### ISSUE TITLE
[Clang] unexpected error: reference to local variable 'X' declared in enclosing function 'Y'

### ISSUE THREAD

**Issue by rupprecht** [llvm/llvm-project#60518]
d708a186b6a9b050d09558163dd353d9f738c82d (reverted in 74ce297045bac4bc475b8e762d2a1ea19bb16d3c) causes an unexpected compilation failure for this:

```c++
struct StringLiteral {
  template <int N>
  StringLiteral(const char (&array)[N])
      __attribute__((enable_if(N > 0 && N == __builtin_strlen(array) + 1,
                               "invalid string literal")));
};

struct Message {
  Message(StringLiteral);
};

void Func1() {
  auto x = Message("x");  // Note: this is fine

  // Note: "xx\0" to force a different type, StringLiteral<3>, otherwise this
  // successfully builds.
  auto y = [&](decltype(Message("xx"))) {};

  // ^ fails with: repro.cc:18:13: error: reference to local variable 'array'
  // declared in enclosing function 'StringLiteral::StringLiteral<3>'

  (void)x;
  (void)y;
}
```

**Comment by cor3ntin** [comment]
That was fixed
